android_app/app/src/main/java/se/lth/math/videoimucapture/IMUManager.java

```diff
@@ -84,10 +90,59 @@ private void setSensorType() {
         else
             ACC_TYPE = Sensor.TYPE_ACCELEROMETER;
         GYRO_TYPE = Sensor.TYPE_GYROSCOPE_UNCALIBRATED;
+        MAG_TYPE = Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED;
+    }
+
+    private float[] linearInterpolate(Deque<SensorPacket> queue, SensorPacket reference) {
+        // target's timestamp is assumed to be within the range of queue's timestamp
+
+        SensorPacket left = null;
+        SensorPacket right = null;
+        Iterator<SensorPacket> itr = queue.iterator();
+
+        // find the closest data right next to gyro's timestamp
+        while (itr.hasNext()) {
+            SensorPacket packet = itr.next();
+
+            // using <= and >= as sometimes there is not enough data & left/right is null
+            if (packet.timestamp <= reference.timestamp) {
+                left = packet;
+            } else if (packet.timestamp >= reference.timestamp) {
+                right = packet;
+                break;
+            }
+        }
+
+        float[] data;
+        if (reference.timestamp - left.timestamp <= mInterpolationTimeResolution) {
+            data = left.values;
+        } else if (right.timestamp - reference.timestamp <= mInterpolationTimeResolution) {
+            data = right.values;
+        } else {
+            float ratio = (float)(reference.timestamp - left.timestamp) /
+                    (right.timestamp - left.timestamp);
+            data = new float[left.values.length]; // could vary depending on sensor type
+            for (int i = 0 ; i < left.values.length ; i++) {
+                data[i] = left.values[i] +
+                        (right.values[i] - left.values[i]) * ratio;
+            }
+        }
+
+        // Remove the current element from the iterator and the list.
+        for (Iterator<SensorPacket> iterator = mAccelData.iterator(); iterator.hasNext(); ) {
```

Comment:
Here the iterator should be over the `queue` instead of `mAccelData`, am I right?

Reply:
right